lib/options.c

```diff
@@ -40,371 +40,249 @@
-{
-    int i;
-    for (i = 0; i < N_TCTI; ++i)
-        if (tcti_type == tcti_map_table[i].type)
-            return tcti_map_table[i].name;
-    return NULL;
-}
-/*
- * Test a common_opts_t structure to be sure the member data has been
- * populated. We don't do any tests on the data for appropriate formats
- * (like testing socket_address for a valid IP address).
- * return 0 if sanity test passes
- * return 1 if help message was explicitly requested
- * return 2 if sanity test fails
- */
-int
-sanity_check_common (common_opts_t  *opts)
-{
-    if (opts->help)
-        return 1;
-    switch (opts->tcti_type) {
-#ifdef HAVE_TCTI_DEV
-    case DEVICE_TCTI:
-        if (opts->device_file == NULL) {
-            fprintf (stderr, "missing --device-file, see --help\n");
-            return 2;
-        }
-        break;
+#ifdef HAVE_TCTI_TABRMD
+#include "tpm2_tools_tcti_abrmd.h"
 #endif
 #ifdef HAVE_TCTI_SOCK
-    case SOCKET_TCTI:
-        if (opts->socket_address == NULL) {
-            fprintf (stderr, "missing --socket-address, see --help\n");
-            return 2;
-        }
-        if (opts->socket_port == 0) {
-            fprintf (stderr, "missing --socket-port, see --help\n");
-            return 2;
-        }
-        break;
+#include "tpm2_tools_tcti_socket.h"
 #endif
-#ifdef HAVE_TCTI_TABRMD
-    case TABRMD_TCTI:
-        /* no options for this one yet */
-        break;
+#ifdef HAVE_TCTI_DEV
+#include "tpm2_tools_tcti_device.h"
 #endif
-    default:
-        fprintf (stderr, "invalid TCTI, see --help\n");
-        return 2;
+
+/* Environment variables usable as alternatives to command line options */
+#define TPM2TOOLS_ENV_TCTI_NAME      "TPM2TOOLS_TCTI_NAME"
+
+struct tpm2_options {
+    tpm2_option_handler on_opt;
+    tpm2_arg_handler on_arg;
+    char *short_opts;
+    size_t len;
+    struct option long_opts[];
+};
+
+tpm2_options *tpm2_options_new(const char *short_opts, size_t len, struct option *long_opts, tpm2_option_handler on_opt, tpm2_arg_handler on_arg) {
+
+    tpm2_options *opts = calloc(1, sizeof(*opts) + (sizeof(*long_opts) * len));
+    if (!opts) {
+        LOG_ERR("oom");
+        return NULL;
     }
-    return 0;
-}
-/*
- * Populate the provided common_opts_t structure with data provided through
- * the environment.
- */
-void
-get_common_opts_from_env (common_opts_t *common_opts)
-{
-    char *env_str, *end_ptr;
-
-    if (common_opts == NULL)
-        return;
-    env_str = getenv (TPM2TOOLS_ENV_TCTI_NAME);
-    if (env_str != NULL)
-        common_opts->tcti_type = tcti_type_from_name (env_str);
-    env_str = getenv (TPM2TOOLS_ENV_DEVICE_FILE);
-    if (env_str != NULL)
-        common_opts->device_file = env_str;
-    env_str = getenv (TPM2TOOLS_ENV_SOCKET_ADDRESS);
-    if (env_str != NULL)
-        common_opts->socket_address = env_str;
-    env_str = getenv (TPM2TOOLS_ENV_SOCKET_PORT);
-    if (env_str != NULL)
-        common_opts->socket_port = strtol (env_str, &end_ptr, 10);
```

Comment:
Are you getting rid of the env vars options for the TCTIs on purpose or did you just forget to parse these in the different tpm2_tools_tcti_*_init() init functions?

I think these are very useful to avoid having to pass the same common TCTI options to each command.

Reply:
They will go to the tcti init routines, where each tcti can handle it's particular options.

Reply:
Agreed with that. I asked because that code was missing (see my others comment in the tcti init funcs).